### DISCUSSION TITLE
Branding → Default Album Art (per station) available via API?

### DISCUSSION THREAD

**Discussion by Moonbase59** [AzuraCast/AzuraCast#6666] [Q&A]
I’m working on an AzuraCast Electronic Program Guide (EPG).

**Is it possible to retrieve the URL to the Default Album Art (from Branding) via the API?**

If not, would be great if we had that like with the streamers API endpoint, so I could include something like a station logo in the EPG.

**Comment by BusterNeece** [comment]
@Moonbase59 

Calling `GET` to `/api/admin/settings/branding` will show you the default album art URL if that's specified, and calling `GET` to `/api/admin/custom_assets/album_art` will show if something's uploaded for custom art.

**Reply by Moonbase59** [reply]
Gee, thanks! My automated EPG and M3U tuner playlist script is almost done. Will be available on Github.

**Reply by Moonbase59** [reply]
@BusterNeece Works like a treat :-) **Is this also possible on a per-station-basis?** (Thinking of people with different settings per station, I use this as an EPG logo.)

**Reply by BusterNeece** [reply]
You'll find the URL in the standard profile edit response: `GET /api/station/(id)/profile/edit`

And you can find the uploaded art in `GET /api/station/(id)/custom_assets/album_art`

**Reply by Moonbase59** [reply]
Cool, thanks for your assistance! This is going to get better and better! For an "EPG station logo", I think I will check the station branding first, then fallback to the admin setting, **would you agree?**

Too bad one needs an API key for getting the image link. This makes it harder for end-users to integrate a station they don’t own into their media center TV/Radio Guide. **Any thoughts on this?**

`azuracast_xmltv` _can_ already do this, but it requires finding and pointing to a logo from a station website, and the commandline gets rather long. Something that might be too much for non-tech-savvy persons.

Do you think I should present a **special `User-agent`** like maybe `azuracast_xmltv/0.3.0`  so station owners know what is accessing their API?

---

This is what I currently get, simply copying your demo instance’s background image link, as an unsuspecting end user might do, and letting `azuracast_xmltv` create the M3U and XML by:

```bash
azuracast_xmltv -u https://demo.azuracast.com -i https://demo.azuracast.com/static/img/hexbg.webp -m
```
![Live-TV – Mozilla Firefox_003](https://github.com/AzuraCast/AzuraCast/assets/3706922/bb7534b4-5a02-443f-bcef-ff376299147c)

There is nothing on schedule in the demo instance, so no entries. @LordHelmchen666 has a nice schedule on their _Eisradio_, though (above yours in the EPG).. I actually listen to that station once in a while, and _love_ that they have a schedule.

_Just imagine: You put a show on schedule, and I can see it on the big TV in my living room (using the Jellyfin AndroidTV client, or KODI), and click on the remote to play. Or change to my own video stream. AzuraCast at its best!_

**Reply by BusterNeece** [reply]
@Moonbase59 So what's the endgame for pulling both the global and per-station custom arts? Because there's an endpoint that basically looks like `/api/nowplaying/(id)/art` and that returns the current album art for the instance, whether it's the current track's or the defaults. We could just as easily build another endpoint that returns the current _default_ album art for the station (by following our standard order-of-operations for what constitutes default album art, at both the station and system level), and spare you a whole bunch of API calls...

**Reply by Moonbase59** [reply]
@BusterNeece Ya know, I was a bit hesitant to ask… Have been flooding you with so many things in the past. Using "album art" to get a station logo isn’t perfect anyway, just a kludge I came up with. 

If a _real_ "station logo" could be had, that could freely be accessed via the API (w/o key) and/or a defined web location (and maybe set in "Branding"), that would allow for much better integration with third-party apps, like radio YP station listings, Electronic Program Guides, and so forth. It _should_ be per station, with a possible overall fallback, I think.

People are visual beings, and the recognition effect is much better using a logo. Just think of all the TV stations.

Any API call saved is a step to more performance and less points of failure, I agree. It’s complicated enough getting all the public/non-public and visible/hidden stations and mount points right.

How do you like the idea of integrating AzuraCast in EPGs in general?

**Reply by BusterNeece** [reply]
I don't mind it, honestly if you're feeling adventurous I'd accept and help work on a pull request to implement this directly rather than having to go through the trouble of building out scripts and cron tasks and API calls and such.

**Reply by Moonbase59** [reply]
Sounds good, only I’m rather out of modern web development using frameworks like Vue and so on. I can read the code and understand much of it, and possibly path things up a bit. In the end, I think I’d like the idea of getting fresh and up-to-date XMLTV-compliant M3U (tuner) playlists and xmltv.xml EPG data.

Then again, almost every media center relies on one or the other "grabbing" software to get their EPG data, especially since TV stations started to want money for this. Jellyfin, KODI, Plex, Emby, TVHeadend, etc. are all alike in that respect, and it seems everyone has agreed on XMLTV as a "standard", using the M3U and XML files. From a distribution, recognition, usability and FOSS perspective, it makes sense to be able to provide AzuraCast schedule data in this form, and give an advantage over others.

If you’re interested, let’s do this step-by-step, I’d suggest:

- Get something going with an "official" station logo.
- Use my Python script first, because it’s been working rather robust for the past weeks, and Python should be universal enough to use it on media servers, even Windows, together with all the other "getting an EPG" modules probably already running there.
- For consolidation for "dumb" servers like Plex (can’t work with many tuner/xmltv files), and personal channel setup, people can use [_xTeVe_](https://github.com/xteve-project/xTeVe), which is just brilliant for setting up personal channel order and filtering out unwanted ones. Plus, it’s a single binary, which eases installation.
- When everything works, `azuracast_xmltv`  is on GitHub, and enough people shown interest, we might consolidate efforts and directly generate the files from within AzuraCast, saving the user installation of one script, and cronjobs. After all, EPG data is relatively seldom pulled, most apps I’ve seen only use an interval of 6, 12 or 24 hours. There _are_ some pitfalls to watch, notably keeping (de-facto) standards and at the same time allow a wide range of clients/servers to auto-assign channels, as an example.
- Coming back to an old feature request,  an EPG should provide content descriptions, so it would be great to be able to fill in the already existing "description" for playlist and streamer schedule blocks to provide that. I think a rather short description might suffice for a radio station, like 256/512 characters or so. (Note: This field allows newlines to display a more attractive description.) Anything to make it nice for station owners and users, and I can avoid hardcoding "Wir spielen die bessere Musik." into each and every description field… ;-) (Or having to read "Playlist: Playlistname" over and over again when it’s already in the title.)

**Comment by BusterNeece** [comment]
So actually, I'm playing with it and this doesn't in fact require the creation of a new API endpoint. You can just misappropriate one of our existing ones!

If you use the "get media artwork" API endpoint and just specify a media ID that outright doesn't (or can't) exist, it'll fall back to showing the station's default artwork, which follows the full order of operations (i.e. preferring per-station uploaded, then per-station URL, then global uploaded, then global URL, then global generic), so basically `/api/station/(id)/art/foo` should yield either an image or a 302 redirect to an image URL. If you want to improve caching on it, you could do `/foo-(unix-timestamp)` and it'll cache that for a year.

**Reply by Moonbase59** [reply]
Hee hee, and I actually _was_ thinking about trying an id of zero, but never actually tried.

I will play with that, but the general idea of having an offical station logo would be worth following, too. Album art could be something very different from the actual logo a station (and it’s website) uses, and logos are often not square.

Thanks for the update!

**Reply by Moonbase59** [reply]
Getting the actual image works fine, but how can I get **the URL** to it?

Reason being: The XMLTV data needs an image _URL_, the actual image is eventually loaded and displayed by the end user's _client_.

**Reply by BusterNeece** [reply]
That API endpoint should be public.

**Reply by Moonbase59** [reply]
It sure is, but I still get an _image_, not a _URL_ to one.

**Reply by BusterNeece** [reply]
The URL to the image is the URL to the API call then... 🤔

**Reply by Moonbase59** [reply]
What are you trying to tell me?

I make the API call. I get an image. I need a URL (in JSON, preferably).

```
 accept-ranges: bytes
 content-length: 380177
 content-type: image/png
 date: Mon,16 Oct 2023 19:48:56 GMT
 etag: "652bcd20-5cd11"
 last-modified: Sun,15 Oct 2023 11:29:36 GMT
 referrer-policy: no-referrer-when-downgrade
 server: openresty
 x-clacks-overhead: GNU Terry Pratchett,Leonard Nimoy
 x-content-type-options: nosniff
 x-firefox-spdy: h2
 x-xss-protection: 1
 ```

Interestingly, I _sent_ an accept `application/json` header.

**Reply by BusterNeece** [reply]
I'm saying it's entirely acceptable to, in your xml file, specify:

```xml
<icon src="https://radio.niteradio.net/api/station/1/art/foo-123.jpg"/>
```

it will either return the image or a 302 to the correct image, but in either case should render perfectly fine

**Reply by Moonbase59** [reply]
Aaaah! Got two teeth pulled today, maybe my brain didn’t start working again with all the medication… :-) Forget the remark.

**Comment by Moonbase59** [comment]
@BusterNeece It is up for you (and everyone) to test:
https://github.com/Moonbase59/azuracast_xmltv

**Comment by Moonbase59** [comment]
@BusterNeece I feel `azuracast_xmltv` is rather feature-complete for what we have with the current AzuraCast. Do you have a suggestion for a more private place where we could continue the discussion, exchange test data and follow up on possible integration strategies?

In the meantime, I’ve been setting up my instance to follow my proposed setup of using
- `https://domain.tld/xmltv/stationshortcode.m3u` for the station M3Us, and
- `https://domain.tld/xmltv/domain.tld.xml` for the EPG data.

**Reply by BusterNeece** [reply]
@Moonbase59 The Discord perhaps? https://discord.gg/azuracast

**Reply by Moonbase59** [reply]
Haven’t used that in ages, but I _should_ have an account somewhere… We’s need to find out our time shift and agree on a time then. I’ll come back on this.
